pycontrails/datalib/ecmwf/model_level_era5.py

```diff
@@ -0,0 +1,501 @@
+        self.download(**xr_kwargs)
+
+        disk_cachepaths = [self.cachestore.get(f) for f in self._cachepaths]
+        ds = self.open_dataset(disk_cachepaths, **xr_kwargs)
+
+        mds = self._process_dataset(ds, **kwargs)
+
+        self.set_metadata(mds)
+        return mds
+
+    @overrides
+    def set_metadata(self, ds: xr.Dataset | MetDataset) -> None:
+        if self.product_type == "reanalysis":
+            product = "reanalysis"
+        elif self.product_type == "ensemble_members":
+            product = "ensemble"
+        else:
+            msg = f"Unknown product type {self.product_type}"
+            raise ValueError(msg)
+
+        ds.attrs.update(
+            provider="ECMWF",
+            dataset="ERA5",
+            product=product,
+        )
+
+    def mars_request(self, times: list[datetime]) -> dict[str, str]:
+        """Generate MARS request for specific list of times.
+
+        Parameters
+        ----------
+        times : list[datetime]
+            Times included in MARS request.
+
+        Returns
+        -------
+        dict[str, str]:
+            MARS request for submission to Copernicus CDS.
+        """
+        unique_dates = set(t.strftime("%Y-%m-%d") for t in times)
+        unique_times = set(t.strftime("%H:%M:%S") for t in times)
+        # param 152 = log surface pressure, needed for metview level conversion
+        grib_params = set(self.variable_ecmwfids + [152])
+        common = {
+            "class": "ea",
+            "date": "/".join(sorted(unique_dates)),
+            "expver": "1",
+            "levelist": "/".join(str(lev) for lev in sorted(self.levels)),
+            "levtype": "ml",
+            "param": "/".join(str(p) for p in sorted(grib_params)),
+            "time": "/".join(sorted(unique_times)),
+            "type": "an",
+            "grid": f"{self.grid}/{self.grid}",
+        }
+        if self.product_type == "reanalysis":
+            specific = {"stream": "oper"}
+        elif self.product_type == "ensemble_members":
+            specific = {"stream": "enda"}
+            if self.ensemble_members is not None:  # always defined; checked to satisfy mypy
+                specific |= {"number": "/".join(str(n) for n in self.ensemble_members)}
+        return common | specific
+
+    def _set_cds(self) -> None:
+        """Set the cdsapi.Client instance."""
+        try:
+            import cdsapi
+        except ModuleNotFoundError as e:
+            dependencies.raise_module_not_found_error(
+                name="ModelLevelERA5._set_cds method",
+                package_name="cdsapi",
+                module_not_found_error=e,
+                pycontrails_optional_package="ecmwf",
+            )
+
+        try:
+            self.cds = cdsapi.Client(url=self.url, key=self.key)
+        # cdsapi throws base-level Exception
+        except Exception as err:
+            raise CDSCredentialsNotFound from err
+
+
+def _download_convert_cache_handler(
```

Comment:
Is there a reason this is outside the class? Is it re-used elsewhere?

Reply:
Nope. I think I was wondering initially if I might be able to re-use this function for the model-level HRES datalib, but they're different enough for that not to make sense. I'll move it inside the class!

Reply:
c4a63967c848ec05fe1c08990ed4c23cf05dba78